### PR TITLE
[gpt_oss_d_p] Top-1 / logits functional sign-off harness (PREFILL_TOP1)

### PR DESCRIPTION
**Pull request by mdragulaTT**
## What

Adds the **top-1 / logits functional sign-off** harness for GPT-OSS prefill — the definitive accuracy
check that KV-cache PCC only proxies. Follow-up to #51992; implements the harness side of #52002.

## Why

Prefill has been validated via **KV-cache PCC** (min ~0.80–0.86 at depth). That's a proxy — it doesn't
answer the real question: **does prefill predict the same tokens as HF at each position?** Top-1 /
logits agreement does.

## This PR

- `runtime.logits_top1_oneshot()` — gathers the SP-seq + TP-vocab sharded prefill logits, drops the
  pow2 vocab zero-padding, argmaxes per position → `[n_tokens]` global token ids. One-shot only
  (single chunk == whole seq); contiguous SP-row order is already natural token order.
- `runtime.top1_check()` — compares to the golden's `reference_top1.safetensors` (key
  `top1_token_ids`) → `(agreement, n_compared, first_token_match)`.
- Harness `PREFILL_TOP1=1` (one-shot): one extra forward with the LM head, compares, and **gates on
  `GPT_OSS_TOP1_MIN`** when set — a CI gate parallel to `GPT_OSS_KV_PCC_MIN`. Absent a reference it
  reports our own prediction so the device path is still exercised.

## Golden dependency

The golden currently has no `reference_top1`. The generator **already runs the full HF forward**
(metadata: `forward_time_seconds`, HF `AutoModelForCausalLM`), so producing it is a ~2-line add:

```python
save_file({"top1_token_ids": logits[0].argmax(-1).to(torch.int32).contiguous()},
          out_dir / "reference_top1.safetensors")  # same token_ids/prompt as the KV golden
```

Once the golden carries it, set `GPT_OSS_TOP1_MIN` and register the galaxy job (#52003).

## Validation

Galaxy 4×8, 36L, real bf16 weights, one-shot 5k: **5000 positions computed**, KV PCC 0.859; device
path OK (reference pending, so agreement is not yet numeric).

## Not in this PR

- Chunked (multi-chunk) top-1 — follow-up.
- The `ATTN_WEIGHT_DTYPE` diagnostic knob — separate follow-up.

Refs: #51992, #52002, #52003

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/gpt-oss-top1-signoff)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/gpt-oss-top1-signoff)

### CODE REVIEW

File: models/demos/gpt_oss_d_p/tt/tt_prefill_runtime.py [hunk 1]
```diff
@@ -307,6 +307,52 @@ def _pcc(a, b):
...
+        if not ref_path.exists():
+            raise FileNotFoundError(
+                f"{ref_path} missing - golden has no reference top-1 yet (ask the golden generator to "
+                f"save logits.argmax(-1) as top1_token_ids)."
+            )
+        with safe_open(str(ref_path), framework="pt") as h:
+            ref = h.get_tensor("top1_token_ids").reshape(-1).long()
+        n = min(ref.numel(), our_top1.numel())
+        agree = float((our_top1[:n].long() == ref[:n]).float().mean())
+        first = bool(our_top1[0].long() == ref[0].long()) if n else False
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/tt_prefill_runtime.py]: Fixed — now compares the argmax at the last prompt position (the first *generated* token); renamed gen_token_match (c37d9d13d0).

File: models/demos/gpt_oss_d_p/tests/galaxy_prefill_kv_pcc.py [hunk 2]
```diff
@@ -200,6 +202,49 @@ def run_once():
...
+                        f"[prefill-pcc] TOP-1 agreement vs golden: {agree * 100:.2f}% over {ncmp} positions "
+                        f"(first-token match={first})",
+                        flush=True,
+                    )
+                    _top1_min = os.environ.get("GPT_OSS_TOP1_MIN")
+                    if _top1_min is not None and agree < float(_top1_min):
+                        print(f"[prefill-pcc] FAIL: TOP-1 {agree:.4f} < GPT_OSS_TOP1_MIN={_top1_min}", flush=True)
+                        return 1
+                except FileNotFoundError as e:
+                    # No reference yet: report our own prediction so the device path is still validated.
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/galaxy_prefill_kv_pcc.py]: Fixed — a missing reference now fails when GPT_OSS_TOP1_MIN is set; only ungated bring-up downgrades to a diagnostic (c37d9d13d0).

File: models/demos/gpt_oss_d_p/tt/tt_prefill_runtime.py [hunk 3]
```diff
@@ -307,6 +307,52 @@ def _pcc(a, b):
...
+
+        ref_path = Path(trace_dir) / "reference_top1.safetensors"
+        if not ref_path.exists():
+            raise FileNotFoundError(
+                f"{ref_path} missing - golden has no reference top-1 yet (ask the golden generator to "
+                f"save logits.argmax(-1) as top1_token_ids)."
+            )
+        with safe_open(str(ref_path), framework="pt") as h:
+            ref = h.get_tensor("top1_token_ids").reshape(-1).long()
+        n = min(ref.numel(), our_top1.numel())
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/tt_prefill_runtime.py]: Fixed — top1_check now rejects a length-mismatched reference (raises) instead of a min() partial compare (c37d9d13d0).

File: models/demos/gpt_oss_d_p/tests/galaxy_prefill_kv_pcc.py [hunk 4]
```diff
@@ -200,6 +202,49 @@ def run_once():
...
             return 1
+
+        # --- functional sign-off: top-1 / logits agreement vs golden (one-shot only) ---
+        # PREFILL_TOP1=1 runs one extra forward with the LM head, gathers the SP-seq + TP-vocab sharded
+        # logits, argmaxes per position, and compares to the golden's reference_top1 (HF argmax). This
+        # is the definitive "does prefill predict the right tokens" check (KV PCC is only a proxy).
+        # Gate on GPT_OSS_TOP1_MIN when set (CI); needs the golden to carry reference_top1.safetensors.
+        if os.getenv("PREFILL_TOP1", "0") == "1":
+            if n_chunks != 1:
+                print("[prefill-pcc] TOP1 needs one-shot (PREFILL_CHUNKED=0); skipping top-1", flush=True)
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/galaxy_prefill_kv_pcc.py]: Fixed — PREFILL_TOP1 on a chunked run now fails (one-shot only) instead of silently passing (c37d9d13d0).

**Comment by mdragulaTT** [comment]
Closing: the golden traces carry KV only, so this harness has nothing to run against, and GPT-OSS prefill work is paused; KV PCC remains the accuracy gate, as for the other prefill models.
